### PR TITLE
Fix handling of new keyword and call signatures and improve type parameter formatting

### PR DESCRIPTION
**Pull request by azz**
* Fix `interface<T> {}`
 * Support `new <T>(): any` and `new <T> => any`
 * Support `<T>() => any` and `<T>(): any`
 * Refactor type parameter formatting to use the same code as flow.
 * Add custom tests for long lines.

### CODE REVIEW

File: tests/typescript/custom/typeParameters/callAndConstructSignatureLong.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+interface Interface {
+    <
+        Voila,
```
Comment (despairblue) [inline tests/typescript/custom/typeParameters/callAndConstructSignatureLong.ts]: Love it 😄

File: src/printer.js [hunk 2]
```diff
@@ -3014,6 +2972,39 @@ function printTypeScriptModifiers(path, options, print) {
   ]);
 }
 
+function printTypeParameters(path, options, print, paramsKey) {
+    const n = path.getValue();
+
+    // In flow, Foo<> is acceptable. In TypeScript, it's a syntax error.
```
Comment (despairblue) [inline src/printer.js]: If it's a syntax error, shouldn't we never get here in the first place?

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,4 @@
 node_modules
 npm-debug.log
 /test.js
+/.vscode
```
Comment (vjeux) [inline .gitignore]: If you could configure your editor to add trailing empty lines that would be great!

File: tests/typescript/conformance/types/constructorType/__snapshots__/jsfmt.spec.js.snap [hunk 4]
```diff
@@ -3,6 +3,6 @@
 exports[`cunstructorType.ts 1`] = `
 var d: new(x: number) => void;
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-var d: new(x: number) => void;
+var d: new (x: number) => void;
```
Comment (vjeux) [inline tests/typescript/conformance/types/constructorType/__snapshots__/jsfmt.spec.js.snap]: is the whitespace here intended?

Comment (azz) [inline tests/typescript/conformance/types/constructorType/__snapshots__/jsfmt.spec.js.snap]: Yes. `new` is a keyword, so the space helps distinguish it from a method.

Comment (vjeux) [inline tests/typescript/conformance/types/constructorType/__snapshots__/jsfmt.spec.js.snap]: Okay, I just wanted to make sure that it wasn't a mistake.

**Pull request merged**